File: toot/step_definitions/Email Marketing Signup.java
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.junit.Assert;
import java.time.Duration;

public class EmailMarketingStepDefinitions {

    WebDriver driver;
    WebDriverWait wait;
    EmailMarketingPage emailMarketingPage;

    public EmailMarketingStepDefinitions() {
        System.setProperty("webdriver.chrome.driver", "path/to/chromedriver");
        driver = new ChromeDriver();
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        emailMarketingPage = new EmailMarketingPage(driver);
    }

    public void givenUserIsOnSignupPage() {
        driver.get("http://example.com/email-signup");
    }

    public void whenUserEntersValidEmail() {
        emailMarketingPage.enterEmail("valid.email@example.com");
    }

    public void whenUserEntersInvalidEmail() {
        emailMarketingPage.enterEmail("invalid-email");
    }

    public void whenUserEntersAlreadyRegisteredEmail() {
        emailMarketingPage.enterEmail("registered.email@example.com");
    }

    public void whenUserSubmitsSignupForm() {
        emailMarketingPage.submitForm();
    }

    public void thenUserShouldSeeConfirmationMessage() {
        Assert.assertTrue("Confirmation message not displayed", emailMarketingPage.isConfirmationMessageDisplayed());
    }

    public void thenUserShouldSeeErrorMessageForInvalidEmail() {
        Assert.assertTrue("Error message for invalid email not displayed", emailMarketingPage.isInvalidEmailErrorMessageDisplayed());
    }

    public void thenUserShouldSeeAlreadyRegisteredMessage() {
        Assert.assertTrue("Already registered message not displayed", emailMarketingPage.isAlreadyRegisteredMessageDisplayed());
    }

    public void thenUserShouldSeeEmailRequiredErrorMessage() {
        Assert.assertTrue("Email required error message not displayed", emailMarketingPage.isEmailRequiredErrorMessageDisplayed());
    }

    public void whenUserClicksPrivacyPolicyLink() {
        emailMarketingPage.clickPrivacyPolicyLink();
    }

    public void thenUserShouldBeRedirectedToPrivacyPolicyPage() {
        Assert.assertTrue("Not redirected to privacy policy page", emailMarketingPage.isPrivacyPolicyPageDisplayed());
    }

    public void thenUserShouldBeAbleToReturnToSignupPage() {
        emailMarketingPage.returnToSignupPage();
        Assert.assertTrue("Not returned to signup page", emailMarketingPage.isSignupPageDisplayed());
    }

    public void tearDown() {
        if (driver != null) {
            driver.quit();
        }
    }
}

File: toot/step_definitions/Email Marketing Signup.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import SignupPage from "../pages/signupPage";

Given("the user is on the email marketing signup page", () => {
 SignupPage.visit();
});

When("the user enters a valid email address", () => {
 SignupPage.enterEmail('valid@example.com');
});

When("the user enters an invalid email address", () => {
 SignupPage.enterEmail('invalid-email');
});

When("the user enters an email address that is already registered", () => {
 SignupPage.enterEmail('registered@example.com');
});

When("the user submits the signup form", () => {
 SignupPage.submitForm();
});

When("the user submits the signup form without entering an email address", () => {
 SignupPage.submitForm();
});

When("the user clicks on the privacy policy link", () => {
 SignupPage.clickPrivacyPolicy();
});

Then("the user should see a confirmation message", () => {
 SignupPage.confirmationMessage().should('be.visible');
});

Then("the user should see an error message indicating the email is invalid", () => {
 SignupPage.errorMessage().should('contain', 'invalid email');
});

Then("the user should see a message indicating the email is already registered", () => {
 SignupPage.alreadyRegisteredMessage().should('contain', 'already registered');
});

Then("the user should see an error message indicating that an email address is required", () => {
 SignupPage.errorMessage().should('contain', 'email address is required');
});

Then("the user should be redirected to the privacy policy page", () => {
 cy.url().should('include', '/privacy-policy');
});

Then("the user should be able to return to the signup page", () => {
 SignupPage.returnToSignupPage();
 cy.url().should('include', '/signup');
});

Then("the user should be added to the email marketing list", () => {
 SignupPage.checkEmailAdded('valid@example.com');
});

Then("the user should not be added to the email marketing list", () => {
 SignupPage.checkEmailNotAdded('invalid-email');
});

Then("the user should not be added to the email marketing list again", () => {
 SignupPage.checkEmailNotAddedAgain('registered@example.com');
});